README.md
<img src="https://raw.githubusercontent.com/appnexus/anx-api/master/docs/img/anx-api.png" alt="AppNexus Api Wrapper" width="300px">

[![npm version](https://badge.fury.io/js/anx-api.svg)](http://badge.fury.io/js/anx-api)
[![Build Status](https://travis-ci.org/appnexus/anx-api.svg?branch=master)](https://travis-ci.org/appnexus/anx-api)
[![Build Dependencies](https://david-dm.org/appnexus/anx-api.png)](https://david-dm.org/appnexus/anx-api)
[![Coverage Status](https://coveralls.io/repos/github/appnexus/anx-api/badge.svg?branch=master)](https://coveralls.io/github/appnexus/anx-api?branch=master)

## Installation

```bash
npm install anx-api --save
```

## Usage Example

```javascript
var AnxApi = require('anx-api');

// Create a new instance with api target
var anxApi = new AnxApi({
    target: 'https://api.appnexus.com'
    token: 'SESSION_TOKEN', // (optional) see also anxApi.login(...)
    rateLimiting: true
});

anxApi.get(<serviceName>).then(function (res) {
    ...
}).catch(function (err) {
    ...
})
```

## Links

* [Upgrading from 3.x to 4.x](docs/upgrading-3.x-to-4.x.md)
* [Upgrading from 2.x to 3.x](docs/upgrading-2.x-to-3.x.md)
* [Changelog](CHANGELOG.md)
* [Contributing](CONTRIBUTING.md)

## Constructor

```javascript
var anxApi = new AnxApi(config);
```

#### Parameters

config[object]:
* `target` - (string) required base api url
* `token` - (string) optional session token
* `request` - (object) optional request object
* `timeout` - (integer) optional request timeout in milliseconds (default: 60000) Triggers `ConnectionAbortedError` error.
* `userAgent` - (string) optional user agent
* `rateLimiting` - (boolean) optional rate limiting
* `concurrencyLimit` - (integer) optional max concurrent requests
* `beforeRequest` - (function) optional before request opts filter (see [beforeRequest](#beforerequest))
* `afterRequest` - (function) optional after request response filter (see [afterRequest](#afterrequest))
* `environment` - (enum:node|browser) optional. node - force axios to run as if in node and not the browser (usefull for testing with jest/jasmine).

## Instance Methods

### #get

Issues a GET request

```javascript
anxApi.get('service url')
anxApi.get('service url', opts)
anxApi.get(opts)
```

#### Parameters

* service uri - (string)
* opts - (object) see [Request Options](#request-options)

#### Response Object

* statusCode - (number) Http response code
* headers - (object) Response headers
* body - (object) Response payload
* req - (object) Request options
* requestTime - (number) Actual on network request time in milliseconds
* totalTime - (number) Total request time including concurrency and rate limiting delays in milliseconds

#### Parameters

* id - (number OR array(number))
* service uri - (string)
* opts - (object) see [Request Options](#request-options)

### #getAll

Usage and parameters are the same as #get accept it pages through api calls.
Response body is parsed as json.

### #post

Issues a POST request with a payload

```javascript
anxApi.post('service url', <payload>)
anxApi.post('service url', <payload>, opts)
anxApi.post(opts)
```

#### Parameters

* service uri - (string)
* payload - (string|object)
* opts - (object) see [Request Options](#request-options)

### #put

Issues a PUT request with a payload

```javascript
anxApi.put('service url', <payload>)
anxApi.put('service url', <payload>, opts)
anxApi.put(opts)
```

#### Parameters

* service uri - (string)
* payload - (string|object)
* opts - (object) see [Request Options](#request-options)


### #delete

Issues a DELETE request

```javascript
anxApi.delete('service url')
anxApi.delete('service url', opts)
anxApi.delete(opts)
```

#### Parameters

* service uri - (string)
* opts - (object) see [Request Options](#request-options)


### #login

Authenticates with the API and returns a token. The token will be reused for future requests.

```javascript
anxApi.login('username', 'password').then(function (token) {

    // The api object is now logged in. Optionally store the token.
    ...

})
```

### #switchUser

```javascript
anxApi.switchUser(userId).then(...)
```

## Request Options

The get, post, put, and delete methods can be called with an opts object. The
opts object has the following request options.

* `uri` - (string) service uri
* `body` - (object) required payload for `.post` and `.put`
* `headers` - (object) optional request header overrides
* `startElement` - (string) optional start index
* `numElements` - (integer) optional number of records to return
* `params` - (object) optional query string parameters
* `mimeType` - (string) optional mimetype
* `timeout` - (integer) optional request timeout in milliseconds (defaults to config timeout)
* `encodeParams` - (boolean) optional whether or not params object should be encoded. (default: false)

### Examples

```javascript
// Fetch the third page of 25 creatives
anxApi.get({
    uri: 'creative',
    startElement: 50,
    numElements: 25
})

anxApi.get('creative', {
    params: {
        start_element: 50,
        num_elements: 25
    }
})

anxApi.get('creative?start_element=50&num_elements=25')
```

# Transforming Request Options and Responses

## beforeRequest

Request options can be modified prior to request execution by supplying a
`beforeRequest` transform function in the constructor. The function should
either return a new options object or `null` which will be ignored.

```javascript
var anxApi = new AnxApi({
    ...,
    beforeRequest: function (opts) {
        var modifiedOpts = _.assign({}, opts, {
            // make changes to options
        })
        return modifiedOpts;
    }
});
```

## afterRequest

Request responses can be modified prior to being delivered by supplying a
`afterRequest` transform function in the constructor. The function should return
either a new response object or `null` which will be ignored.

```javascript
var anxApi = new AnxApi({
    ...,
    afterRequest: function (res) {
        var modifiedRes = _.assign({}, res, {
            // make changes to response
        })
        return modifiedRes;
    }
});
```

# Error Handling

```javascript
anxApi.get('creative').then(function (res) {
    ...
}).catch(function (err) {
    if (err instanceof NotAuthenticatedError) {
        console.log('Your not logged in!');
    }
})
```

## Error Object Types

* `Error` - generic error type
* `NetworkError`
    * `DNSLookupError` - target host could not be looked up [ENOTFOUND]
    * `ConnectionAbortedError` - request timeout was reached [ECONNABORTED]
    * `ConnectionRefusedError` - [ECONNREFUSED]
    * `ConnectionResetError` - [ECONNRESET]
    * `ConnectionTimeoutError` - [ETIMEDOUT]
    * `SocketTimeoutError` - [ESOCKETTIMEDOUT]
* `ApiError` - base api error type
    * `NotAuthenticatedError` - token is invalid or expired
    * `NotAuthorizedError` - Unauthorized to make request
    * `RateLimitExceededError`
    * `SystemServiceUnavailableError`
    * `SystemUnknownError`
    * `TargetError` - target was not supplied

## Error Object Properties

* id - Api error id
* code - Api error code
* message - Brief error description
* req - Request options used in failed request
* res - Raw failed response

# Custom Request and Debugging

The following are two different methods of modifying and or spying on requests
made to the api.

### Wrap the internal request function

```javascript
anxApi._config.request = _.wrap(anxApi._config.request, function (request, opts) {
    console.log('DEBUG: ', opts);
    return request.call(api, opts);
});
```

### Pass in a custom request object

```javascript
var request = require('request');

function customRequest(opts) {
    return new Promise(function (resolve, reject) {

        // Customize the request

        request(opts, function (err, res) {
            if (err) {

                // Add additional error handling

                reject(err);
            } else {

                // Customize the response

                resolve(res);
            }
        });
    });
}

var anxApi = new AnxApi({
    target: process.env.ANX_TARGET,
    token: 'SESSION_TOKEN',
    request: customRequest
});
```

## Tests

### Running unit tests

Run the unit test suite from the project root, make sure you've run `npm
install` first:

```bash
npm test
```

## License

See [LICENSE](LICENSE) file


CHANGELOG.md
# Change Log

## v8.0.2
- [patch] Updated axios
- [patch] Removed pre-push for security reasons
- [patch] Updated dev dependencies

## v8.0.1
- [patch] Updated axios
- [patch] Updated dev dependencies

## v8.0.0

- [major] Reworked errors
- [patch] Added eslint
- [patch] Various lint fixes

## v7.1.1

- [patch] Bumped jest and ts-jest versions

## v7.1.0

- [patch] More strict internal types
- [patch] Unit tested concurrency
- [minor] Added `statusCode` and `uri` to the root response object

## v7.0.0

- [major] Converted project to typescript

## v6.1.0

- [minor] Updated dependencies and converted from nsp to npm audit

## v6.0.0

[major] Adds in the package `url-join` to properly concat URL's when building the request. This breaking change requires partial url segments to be used when making requests instead of specifying fully-qualified paths. See: https://github.com/appnexus/anx-api/issues/70

## v5.0.4

- [patch] reverts changes made in v5.0.3 due to breaking change

## v5.0.3

- [patch] Adds in the package `url-join` to properly concat URL's when building the request

## v5.0.2

- [patch] Removes experimental `getBy` and `getById` methods.

## v5.0.1

- [minor] Adds `.npmignore` file to fix npm published version.

## v5.0.0

- [major] Removed use of `es6-promise` library; native promises are used now.
- [minor] Adds `lodash-babel-plugin` to optimize size for browser use case.

## v4.4.1

- Fixed rate limiting to abort if retry-after header is missing

## v4.4.0

- [minor] Added optional `environment` config option

## v4.3.0

- [minor] Added `encodeParams` option for params

## v4.2.0

- [minor] Added original request options to the response `res.req`
- Converted from Mocha to Jest
- Added eslint-config-appnexus
- Fixed missing content in request queue
- Added coveralls.io
- Upgraded nock

## v4.1.0

- Fixed rate limiting
- Updated Axios@0.15.0
- 'DNSLookupError' now inherits from 'NetworkError'
- [minor] Added `NetworkError`, with `ConnectionAbortedError`, `SocketTimeoutError`, `ConnectionTimeoutError`, `ConnectionRefusedError` sub classes.
- Fixed undefined err object

## v4.0.1

- Fixed bug in error response

## v4.0.0

- [minor] Added experimental `getBy` method
- [minor] Added request options in the response object `res.req`
- [minor] Added request options in the error object `err.req`
- [minor] Added request response in the error object `err.res`
- [major] Revamped internal error building, error api slightly changed
- Updated dependencies
- [minor] Added totalTime to response object
- [major] Removed deprecated methods `requestJson`, `getJson`, `getAllJson`, `postJson`, `putJson`, and `deleteJson`

## v3.3.1

- Fixed bug in `getAll` where it did not fail on first api error
- Fixed formatting for DNS Lookup Errors
- Fixed bug in axiosAdapter error handling
- Added pre-push lint and dependency checks

## v3.3.0

- [minor] Added default timeout to 60 seconds
- [minor] Added experimental `getById` method
- Added ability to flag methods as experimental
- Refactored `request` to eliminate unneeded function
- Refactored `getAll` and added unit tests for it

## v3.2.1

- Updated dependencies
- [minor] Added config unit test
- Fixed error handling for newer axios version

## v3.2.0

- [minor] Rate limiting on by default
- [minor] Added support for retry-after header [API v1.17]
- Added rate limit detection with status code 429 [API v1.17]

## v3.1.2

- Deprecated `getAllJson` in favor of `getAll`
- 'getAllJson' no longer designated as experimental.
- Fixed paging in `getAllJson`

## v3.1.1

- Temporary fix for FireFox not supporting `Error.captureStackTrace`
- Fixed various bugs with error handling.
- Added `DNSLookupError`

## v3.1.0

- [minor] Added `concurrencyLimit` option to the constructor
- Updated `qs`

## v3.0.0

- Added `rateLimiting` option to the constructor
- Added `mimeType` option to requests
- Setting `opts.headers.Accept` and `opts.headers['Content-Type']` overrides json defaults
- Fixed bug where `get` and `delete` set `Content-Type`
- Replaced `q` with `es6-promise`
- Added error type `RateLimitExceededError`
- Added `afterRequest` config function to transform response objects
- Replaced `gulp` with npm scripts
- Replaced `jshint` for `eslint`
- Methods now allow the format .method(url[string], opts[object]);
- Added error types `SystemServiceUnavailableError` and `SystemServiceUnavailableError`
- `getJson`, `postJson`, `putJson`, and `deleteJson` are deprecated
- Replaced `request` with `axios` adapter to make `anx-api` isomorphic

## v2.2.0

- Updated `qs` and `lodash`
- Allows null or undefined urls
- Allows un-setting 'User-Agent' with config.userAgent = null
- Added `beforeRequest` config function to transform request options
- Added ability to set default headers in the config.
- Added change log

## v2.x

- Methods return promises and callbacks are removed.


lib/api.mspec.js
/* eslint func-names: 0, padded-blocks: 0 */

var _ = require('lodash');
var sinon = require('sinon');
var assert = require('assert');
var Promise = require('es6-promise').Promise;

var AnxApi = require('./api');

describe('AnxApi', function() {

	describe('Request', function() {

		describe('config', function() {
			var opts;
			var res;

			describe('with valid config', function() {

				beforeEach(function() {
					opts = null;
					res = null;

					var api = new AnxApi({
						target: 'http://example.com',
						token: 'MySessionToken',
						userAgent: 'MyAgent',
						rateLimiting: false,
						request: function(o) {
							opts = o;
							return Promise.resolve({ testKey: 'testValue' });
						},
						beforeRequest: function(reqOpts) {
							return _.assign({}, reqOpts, { body: 'test' });
						},
						afterRequest: function(reqRes) {
							return _.assign({}, reqRes, { afterRequest: 'value' });
						},
					});

					return api.get('/user').then(function(reqRes) {
						res = reqRes;
						return null;
					});
				});

				it('should use target', function() {
					assert(_.includes(opts.uri, 'http://example.com/'));
				});

				it('should use token', function() {
					assert.equal(opts.headers.Authorization, 'MySessionToken');
				});

				it('should use User-Agent', function() {
					assert.equal(opts.headers['User-Agent'], 'MyAgent');
				});

				it('should use beforeRequest function', function() {
					assert.equal(opts.body, 'test');
				});

				it('should use afterRequest function', function() {
					assert.deepEqual(res.afterRequest, 'value');
				});

				it('should default request timeout', function() {
					assert.equal(opts.timeout, 60000);
				});

			});

			describe('with invalid config', function() {

				it('should throw on missing target', function(done) {
					var api = new AnxApi({
						rateLimiting: false,
					});

					api.get('/user').catch(function(err) {
						assert(err instanceof AnxApi.TargetError);
						done();
					});
				});

			});

		});

		describe('Options', function() {
			var opts;

			beforeEach(function(done) {
				opts = null;
				var api = new AnxApi({
					target: 'http://example.com',
					rateLimiting: false,
					request: function(o) {
						opts = o;
						done();
					},
				});
				api.get({
					uri: 'user',
					timeout: 5000,
					startElement: 100,
					numElements: 25,
					params: {
						myParam: 'value',
						myStdArray: [
							1,
							2,
							3,
						],
						myObjArray: [{
							a: 'apple',
						}, {
							b: 'bee',
						}],
					},
				});
			});

			it('uri should contain start_element', function() {
				assert(_.includes(opts.uri, 'start_element=100'));
			});

			it('uri should contain num_elements', function() {
				assert(_.includes(opts.uri, 'num_elements=25'));
			});

			it('uri should contain params', function() {
				assert(_.includes(opts.uri, 'myParam=value'));
			});

			it('uri should convert standard nested array params', function() {
				assert(_.includes(opts.uri, 'myStdArray[0]=1&myStdArray[1]=2&myStdArray[2]=3'));
			});

			it('uri should convert nested object array params', function() {
				assert(_.includes(opts.uri, 'myObjArray[0][a]=apple&myObjArray[1][b]=bee'));
			});

			it('should default request timeout', function() {
				assert.equal(opts.timeout, 5000);
			});

			describe('validation', function() {
				var api;
				var reqOpts;

				beforeEach(function() {
					api = new AnxApi({
						target: 'http://example.com',
						rateLimiting: false,
						request: function(o) {
							reqOpts = o;
							return {
								then: function(callback) {
									callback({});
								},
							};
						},
					});
				});

				function expectValidationError(errOpts, expected, done) {
					api.request(errOpts).then(function() {
						return done(new Error('Expected error: ' + expected));
					}).catch(function(err) {
						assert(err instanceof AnxApi.ArgumentError);
						if (err.message !== expected) {
							return done(new Error('Unexpected error message: ' + err.message + ' Expected: ' + expected));
						}
						return done();
					});
				}

				_.each({
					uri: [
						{ value: 'creative' },
						{ value: null },
						{ value: undefined },
					],
					startElement: [
						{ value: null },
						{ value: undefined },
						{ value: 0 },
						{ value: 1.1, message: 'Invalid startElement' },
						{ value: '', message: 'Invalid startElement' },
						{ value: 5, uriContains: 'start_element=5' },
						{ value: '10', uriContains: 'start_element=10' },
						{ value: 'ZZZ', message: 'Invalid startElement' },
					],
					numElements: [
						{ value: null },
						{ value: undefined },
						{ value: 0 },
						{ value: 1.1, message: 'Invalid numElements' },
						{ value: '', message: 'Invalid numElements' },
						{ value: 5, uriContains: 'num_elements=5' },
						{ value: '10', uriContains: 'num_elements=10' },
						{ value: 'ZZZ', message: 'Invalid numElements' },
					],
				}, function(tests, param) {
					describe(param, function() {
						tests.forEach(function(test) {
							var newOpts = {};
							if (param !== 'uri') {
								newOpts.uri = '/user';
							}
							newOpts[param] = test.value;
							if (test.message) {
								it(param + ' should not accept ' + test.value, function(done) {
									expectValidationError(newOpts, test.message, done);
								});
							} else {
								it(param + ' should accept ' + test.value, function() {
									return api.request(newOpts).then(function() {
										if (test.uriContains) {
											assert(_.includes(reqOpts.uri, test.uriContains), 'Url (' + reqOpts.uri + ') did not contain: ' + test.uriContains);
										}
										return null;
									});
								});
							}
						});
					});
				});

			});

			describe('Errors', function() {

				describe('NotAuthenticatedError', function() {
					var api;

					beforeEach(function() {
						api = new AnxApi({
							target: 'http://example.com',
							userAgent: 'MyAgent',
							rateLimiting: false,
							request: function() {
								return {
									then: function(callback) {
										callback({
											statusCode: 401,
										});
									},
								};
							},
						});
					});

					it('should reject with NotAuthenticatedError', function(done) {
						api.get('user').then(function() {
							return done(new Error('Did not catch 401'));
						}).catch(function(err) {
							if (!(err instanceof AnxApi.NotAuthenticatedError)) {
								return done(new Error('Did not catch NotAuthenticatedError'));
							}
							return done();
						});
					});

				});

			});

		});

	});

	describe('#request', function() {

		describe('opts.headers', function() {
			var api;

			beforeEach(function() {
				api = new AnxApi({
					target: 'http://example.com',
					rateLimiting: false,
					request: function(opts) {
						return Promise.resolve(opts);
					},
				});
			});

			describe('json default', function() {

				it('should set up GET request for json', function() {
					return api.request({}).then(function(opts) {
						assert.equal(opts.headers.Accept, 'application/json', 'bad or missing Accept');
						assert(!opts.headers['Content-Type'], 'Content-Type should not be set');
						return null;
					});
				});

				it('should set up POST request for json', function() {
					return api.request({ method: 'POST' }).then(function(opts) {
						assert.equal(opts.headers.Accept, 'application/json', 'bad or missing Accept');
						assert.equal(opts.headers['Content-Type'], 'application/json', 'bad or missing Content-Type');
						return null;
					});
				});

				it('should set up PUT request for json', function() {
					return api.request({ method: 'PUT' }).then(function(opts) {
						assert.equal(opts.headers.Accept, 'application/json', 'bad or missing Accept');
						assert.equal(opts.headers['Content-Type'], 'application/json', 'bad or missing Content-Type');
						return null;
					});
				});

				describe('header overrides', function() {
					it('should allow overrideing Accept', function() {
						return api.request({ method: 'DELETE' }).then(function(opts) {
							assert.equal(opts.headers.Accept, 'application/json', 'bad or missing Accept');
							assert(!opts.headers['Content-Type'], 'Content-Type should not be set');
							return null;
						});
					});

					it('should allow overriding Content-Type', function() {
						return api.request({ method: 'GET', headers: { 'Content-Type': 'application/json' } }).then(function(opts) {
							assert.equal(opts.headers.Accept, 'application/json', 'bad or missing Accept');
							assert(opts.headers['Content-Type'], 'Content-Type should be set');
							return null;
						});
					});
				});

			});

			it('should allow overriding json accept type', function() {
				return api.request({ method: 'POST', headers: { Accept: 'text/csv', 'Content-Type': 'text/csv' }}).then(function(opts) {
					assert.equal(opts.headers.Accept, 'text/csv', 'bad or missing Accept');
					assert.equal(opts.headers['Content-Type'], 'text/csv', 'bad or missing Content-Type');
					return null;
				});
			});

			it('should allow setting Accept and Content-Type with mimeType option', function() {
				return api.request({ method: 'POST', mimeType: 'text/csv' }).then(function(opts) {
					assert.equal(opts.headers.Accept, 'text/csv', 'bad or missing Accept');
					assert.equal(opts.headers['Content-Type'], 'text/csv', 'bad or missing Content-Type');
					return null;
				});
			});

		});

	});

	describe('#get', function() {

		describe('opts', function() {
			var opts;

			beforeEach(function(done) {
				opts = null;
				var api = new AnxApi({
					target: 'http://example.com',
					rateLimiting: false,
					request: function(o) {
						opts = o;
						done();
					},
				});
				api.get('user');
			});

			it('method should be GET', function() {
				assert.equal(opts.method, 'GET');
			});

			it('should use string path', function() {
				assert(_.includes(opts.uri, 'http://example.com/user'));
			});

		});

	});

	describe('#getById', function() {
		var api;
		var requestStub;

		beforeEach(function() {
			requestStub = sinon.stub();
			api = new AnxApi({
				target: 'http://example.com',
				request: requestStub,
				rateLimiting: false,
				chunkSize: 2,
			});
		});

		it('should get single id', function() {
			requestStub.resolves({ body: { response: { count: 1, user: { id: 3 }, dbg_info: { output_term: 'user' } } } });
			return api.getById(3, 'test').then(function(res) {
				assert.deepEqual(requestStub.firstCall.args[0].params, { id: 3 });
				assert.deepEqual(res.body, {
					response: {
						count: 1,
						dbg_info: {
							output_term: 'user',
						},
						user: {
							id: 3,
						},
					},
				});
				return null;
			});
		});

		it('should get multi ids', function() {
			requestStub.onCall(0).resolves({ body: { response: { users: [{ id: 1 }, { id: 2 }], dbg_info: { output_term: 'users' } } } });
			requestStub.onCall(1).resolves({ body: { response: { user: { id: 3 }, dbg_info: { output_term: 'user' } } } });
			return api.getById([1, 2, 3], 'test').then(function(res) {
				assert.deepEqual(requestStub.firstCall.args[0].params, { id: '1,2' });
				assert.deepEqual(requestStub.secondCall.args[0].params, { id: '3' });
				assert.deepEqual(res.body, {
					response: {
						count: 3,
						dbg_info: {
							output_term: 'users',
						},
						users: [{
							id: 1,
						}, {
							id: 2,
						}, {
							id: 3,
						}],
					},
				});
				return null;
			});
		});

		it('should reject invalid ids', function(done) {
			api.getById([], 'test').then(function() {
				return done(new Error('ArgumentError not detected'));
			}).catch(function(err) {
				assert(err instanceof AnxApi.ArgumentError);
				done();
			});
		});

	});

	describe('#getBy', function() {
		var api;
		var requestStub;

		beforeEach(function() {
			requestStub = sinon.stub();
			api = new AnxApi({
				target: 'http://example.com',
				request: requestStub,
				rateLimiting: false,
				chunkSize: 2,
			});
		});

		it('should get single param', function() {
			requestStub.resolves({ body: { response: { count: 1, user: { name: 3 }, dbg_info: { output_term: 'user' } } } });
			return api.getBy('name', 3, 'test').then(function(res) {
				assert.deepEqual(requestStub.firstCall.args[0].params, { name: 3 });
				assert.deepEqual(res.body, {
					response: {
						count: 1,
						dbg_info: {
							output_term: 'user',
						},
						user: {
							name: 3,
						},
					},
				});
				return null;
			});
		});

		it('should get multi ids', function() {
			requestStub.onCall(0).resolves({ body: { response: { users: [{ id: 1 }, { id: 2 }], dbg_info: { output_term: 'users' } } } });
			requestStub.onCall(1).resolves({ body: { response: { user: { id: 3 }, dbg_info: { output_term: 'user' } } } });
			return api.getBy('name', [1, 2, 3], 'test').then(function(res) {
				assert.deepEqual(requestStub.firstCall.args[0].params, { name: '1,2' });
				assert.deepEqual(requestStub.secondCall.args[0].params, { name: '3' });
				assert.deepEqual(res.body, {
					response: {
						count: 3,
						dbg_info: {
							output_term: 'users',
						},
						users: [{
							id: 1,
						}, {
							id: 2,
						}, {
							id: 3,
						}],
					},
				});
				return null;
			});
		});

		it('should reject invalid ids', function(done) {
			api.getBy('id', [], 'test').then(function() {
				return done(new Error('ArgumentError not detected'));
			}).catch(function(err) {
				assert(err instanceof AnxApi.ArgumentError);
				done();
			});
		});

	});

	describe('#getAll', function() {
		var api;
		var requestStub;

		beforeEach(function() {
			requestStub = sinon.stub();
			api = new AnxApi({
				target: 'http://example.com',
				request: requestStub,
				rateLimiting: false,
			});
		});

		it('should ', function() {
			requestStub.onCall(0).resolves({ body: { response: { status: 'OK', count: 3, num_elements: 2, users: [{ id: 1 }, { id: 2 }], dbg_info: { output_term: 'users' } } } });
			requestStub.onCall(1).resolves({ body: { response: { status: 'OK', count: 3, num_elements: 2, user: { id: 3 }, dbg_info: { output_term: 'user' } } } });
			return api.getAll('user').then(function(res) {
				assert.equal(requestStub.firstCall.args[0].uri, 'http://example.com/user?start_element=0&num_elements=100');
				assert.equal(requestStub.secondCall.args[0].uri, 'http://example.com/user?start_element=2&num_elements=2');
				assert.deepEqual(res.body, {
					response: {
						count: 3,
						dbg_info: {
							output_term: 'users',
							time: 0,
						},
						num_elements: 3,
						start_element: 0,
						users: [{
							id: 1,
						}, {
							id: 2,
						}, {
							id: 3,
						}],
					},
				});
				return null;
			});
		});

	});

	describe('#post', function() {

		describe('opts', function() {
			var opts;

			beforeEach(function(done) {
				var api = new AnxApi({
					target: 'http://example.com',
					rateLimiting: false,
					request: function(o) {
						opts = o;
						done();
					},
				});
				api.post('user', { name: 'MyName' });
			});

			it('method should be POST', function() {
				assert.equal(opts.method, 'POST');
			});

			it('should use string path', function() {
				assert(_.includes(opts.uri, 'http://example.com/user'));
			});

			it('should place the payload into the post body', function() {
				assert.deepEqual(opts.body, { name: 'MyName' });
			});

		});

	});

	describe('#put', function() {

		describe('opts', function() {
			var opts;

			beforeEach(function(done) {
				var api = new AnxApi({
					target: 'http://example.com',
					rateLimiting: false,
					request: function(o) {
						opts = o;
						done();
					},
				});
				api.put('user', { name: 'MyName' });
			});

			it('method should be PUT', function() {
				assert.equal(opts.method, 'PUT');
			});

			it('should use string path', function() {
				assert(_.includes(opts.uri, 'http://example.com/user'));
			});

			it('should place the payload into the put body', function() {
				assert.deepEqual(opts.body, { name: 'MyName' });
			});

		});

	});

	describe('#delete', function() {

		describe('opts', function() {
			var opts;

			beforeEach(function(done) {
				var api = new AnxApi({
					target: 'http://example.com',
					rateLimiting: false,
					request: function(o) {
						opts = o;
						done();
					},
				});
				api.delete('user?id=1');
			});

			it('method should be DELETE', function() {
				assert.equal(opts.method, 'DELETE');
			});

			it('should use string path', function() {
				assert(_.includes(opts.uri, 'http://example.com/user'));
			});

		});

	});

	describe('#statusOk', function() {

		it('should return true when status is OK', function() {
			assert.equal(AnxApi.statusOk({ response: { status: 'OK' } }), true);
		});

		it('should return false when status is not OK', function() {
			assert.equal(AnxApi.statusOk({ response: { status: '' } }), false);
		});

		it('should return false with no status field', function() {
			assert.equal(AnxApi.statusOk({ response: {} }), false);
		});

		it('should return false with no response field', function() {
			assert.equal(AnxApi.statusOk({}), false);
		});

	});

	describe('#login', function() {

		function buildApi(responseData) {
			var api = new AnxApi({
				// target: 'http://example.com',
				target: 'https://sand.api.appnexus.com',
				userAgent: 'MyAgent',
				rateLimiting: false,
				request: function() {
					return new Promise(function(resolve) {
						resolve(responseData);
					});
				},
			});
			return api;
		}

		it('should reject with NotAuthenticatedError if status is not ok', function() {
			var api = buildApi({
				statusCode: 200, // 200 instead of 401 because thats what the service responds with
				body: {
					response: {
						error_id: 'UNAUTH',
						error: 'No match found for user/pass',
						error_description: null,
						service: null,
						method: null,
						error_code: null,
					},
				},
			});
			return api.login('test_user', 'bad_password').then(function() {
				throw new Error('Did not catch Login Error');
			}).catch(function(err) {
				// API treats bad apssword as Authentication instead of Authorization Error.
				// assert(err instanceof AnxApi.NotAuthenticatedError, 'Api.NotAuthenticatedError');
				assert(err instanceof AnxApi.NotAuthorizedError, 'Api.NotAuthorizedError');

				assert.equal('UNAUTH', err.id);
				assert.equal('No match found for user/pass', err.message);
			});
		});

		it('should login give api auth token', function() {
			var api = buildApi({
				statusCode: 200,
				body: {
					response: {
						status: 'OK',
						token: 'hbapi:10340:55ba41134f752:lax1',
					},
				},
			});
			return api.login('test_user', 'test_password').then(function(token) {
				assert.equal('hbapi:10340:55ba41134f752:lax1', token);
				return null;
			});
		});
	});

	describe('#switchUser', function() {
		var api;
		var requestStub;

		beforeEach(function() {
			requestStub = sinon.stub();
			api = new AnxApi({
				target: 'http://example.com',
				request: requestStub,
				rateLimiting: false,
			});
		});

		it('should post to /auth', function() {
			requestStub.resolves({});
			return api.switchUser(1234).then(function() {
				assert.equal(requestStub.firstCall.args[0].method, 'POST');
				assert.equal(requestStub.firstCall.args[0].uri, 'http://example.com/auth');
				assert.deepEqual(requestStub.firstCall.args[0].body, { auth: { switch_to_user: 1234 }});
				return null;
			});
		});

	});

});
